src/charm.py

```diff
@@ -18,3 +28,65 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
+        self._charm_state = None
+        self._saml_integrator = None
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.saml_relation_created, self._on_saml_relation_created)
+
+    def _on_upgrade_charm(self, _) -> None:
+        """Install needed apt packages."""
+        self.unit.status = ops.MaintenanceStatus("Installing packages")
+        apt.update()
+        apt.add_package(["libxml2"])
+
+    def _on_config_changed(self, _) -> None:
+        """Handle changes in configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring charm")
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+            self._saml_integrator = SamlIntegrator(charm_state=self._charm_state)
```

Comment:
This should also be initialized in the constructor

Reply:
Why's that? I wan't to reinstantiate on every config changed event

Reply:
Charms are stateless, they are re-instantiated for every hook (unless internal events, and defered events)

Reply:
That would work with a deferred config changed event, then. Am I right?

Reply:
Nope but the outcome is "Do not use defer"

Reply:
The testing framework doesn't reinstantiate, so I suggest we keep it as it is until it gets fixed